src/kernel/include/pgtbl.h

```diff
@@ -57,245 +70,398 @@ struct tlb_quiescence {
+//
+//	return 0;
+//}
+//
+//static inline void *__pgtbl_getleaf(struct ert_intern *a, void *accum)
+//{ if (unlikely(!a)) return NULL; return __pgtbl_get(a, accum, 1); }
 /*
- * Use the passed in page, but make sure that we only use the passed
- * in page once.
- */
-static inline void *
-__pgtbl_a(void *d, int sz, int leaf)
-{
-	void **i = d, *p;
-
-	(void)leaf;
-	assert(sz == PAGE_SIZE);
-	if (unlikely(!*i)) return NULL;
-	p = *i;
-	*i = NULL;
-	return p;
-}
-static struct ert_intern *
-__pgtbl_get(struct ert_intern *a, void *accum, int isleaf)
-{
-	(void)isleaf;
-	/* don't use | here as we only want the pte flags */
-	*(u32_t*)accum = (((u32_t)a->next) & PGTBL_FLAG_MASK);
-	return chal_pa2va((paddr_t)((((u32_t)a->next) & PGTBL_FRAME_MASK)));
-}
-static int __pgtbl_isnull(struct ert_intern *a, void *accum, int isleaf)
-{ (void)isleaf; (void)accum; return !(((u32_t)(a->next)) & (PGTBL_PRESENT|PGTBL_COSFRAME)); }
-static void
-__pgtbl_init(struct ert_intern *a, int isleaf)
-{
-	(void)isleaf;
-//	if (isleaf) return;
-	a->next = NULL;
-}
-
-/* We only need to do mapping_add at boot time to add all physical
- * memory to the pgtbl of llboot. After that, we only need to do copy
- * from llboot pgtbl to other pgtbls. Thus, when adding to pgtbl, we
- * use physical addresses; when doing copy, we don't need to worry
- * about PA. */
-
-/* v should include the desired flags */
-static inline int
-__pgtbl_setleaf(struct ert_intern *a, void *v)
-{
-	u32_t new, old;
-
-	old = (u32_t)(a->next);
-	new = (u32_t)(v);
-
-	if (!cos_cas((unsigned long *)a, old, new)) return -ECASFAIL;
-
-	return 0;
-}
-
-/* This takes an input parameter as the old value of the mapping. Only
- * update when the existing value matches. */
-static inline int
-__pgtbl_update_leaf(struct ert_intern *a, void *v, u32_t old)
-{
-	u32_t new;
-
-	new = (u32_t)(v);
-	if (!cos_cas((unsigned long *)a, old, new)) return -ECASFAIL;
-
-	return 0;
-}
-
-/* Note:  We're just using pre-defined default flags for internal (pgd) entries */
-static int
-__pgtbl_set(struct ert_intern *a, void *v, void *accum, int isleaf)
-{
-	u32_t old, new;
-	(void)accum; assert(!isleaf);
-
-	old = (u32_t)a->next;
-	new = (u32_t)chal_va2pa((void*)((u32_t)v & PGTBL_FRAME_MASK)) | PGTBL_INTERN_DEF;
-
-	if (!cos_cas((unsigned long *)&a->next, old, new)) return -ECASFAIL;
-
-	return 0;
-}
-
-static inline void *__pgtbl_getleaf(struct ert_intern *a, void *accum)
-{ if (unlikely(!a)) return NULL; return __pgtbl_get(a, accum, 1); }
-
 ERT_CREATE(__pgtbl, pgtbl, PGTBL_DEPTH, PGTBL_ORD, sizeof(int*), PGTBL_ORD, sizeof(int*), NULL, \
 	   __pgtbl_init, __pgtbl_get, __pgtbl_isnull, __pgtbl_set,	\
 	   __pgtbl_a, __pgtbl_setleaf, __pgtbl_getleaf, ert_defresolve);
+*/
+/* How many bits have been translated ? */
+/*
+first layer, size_order 2^32, 5 bits to represent. However we only allow 8-pgtbls, thus, the last layer is fixed to 8.
+
+entry_order? ignore, because entry_order is really sucky.
```

Comment:
this isn't a very coherent comment.  Please take more time in writing out what you want in the comment.  You need one here, but you need to make sure that it is written well so that people can follow.